### ISSUE TITLE
[feature request] scan QR code from selected region

### ISSUE THREAD

**Issue by cdluminate** [lupoDharkael/flameshot#511]
It would be better if flameshot can scan QR code from the selected region, and put the decoded text to clipboard.

**Comment by cdluminate** [comment]
Code reference: https://launchpad.net/qr-tools

**Comment by ZetaoYang** [comment]
@cdluminate Your idea is good, I will achieve it. Thanks again.

**Comment by data-man** [comment]
I think [quirc](https://github.com/dlbeer/quirc) is better. This library included to OpenCV.

**Comment by ZetaoYang** [comment]
This could be realized by `opencv`  [detectAndDecode](https://docs.opencv.org/3.4/de/dc3/classcv_1_1QRCodeDetector.html#a3ae501d2e9061d6122ae59c005d05edc).

**Comment by greno4ka** [comment]
Flameshot is a screenshot tool. Adding QR grabber to it will look like making knife and flashlight in one tool. IMHO, that's not a good idea. Better to concentrate attention on actual bugs and issues.

**Comment by martianzhang** [comment]
Ubuntu 20.04 can do this by following steps.

```bash
# install zbar-tools
sudo apt install zbar-tools

# qr-scan script
cat > qr-scan
#!/bin/bash

flameshot gui -r | zbarimg - | sed "s/^QR-Code://g" | xclip -selection clipboard 2>/dev/null
Ctrl+D

# add keyboard shortcut for qr-scan script
```

**Comment by xstable** [comment]
@martianzhang I had modified your suggestion to be able to analyse any still taken image (which inside xclip) for qr-code:
```
#!/bin/bash
xclip -selection clipboard -t image/jpeg -o | zbarimg - | sed "s/^QR-Code://g" | xclip -selection clipboard 2>/dev/null 

```
In my case, I do add it straight to the Klipper (KDE) as action:
![image](https://user-images.githubusercontent.com/1778068/148922958-f0cac4fe-f2cd-48cd-9110-fb71c0bb98d3.png)

![image](https://user-images.githubusercontent.com/1778068/148923567-aca3a0c0-4863-4879-af69-d54c1888bc3a.png)

**Comment by data-man** [comment]
I think `zbarimg - | sed "s/^QR-Code://g"` can be replaced to `zbarimg --raw -q`.

**Comment by mmahmoudian** [comment]
@xstable 
> Alternatively it might be a good feature to have something like a queue, which pipe the image we create to every script we like.
In this case we simply could create an entry "qr-codeScanner" and put the zbarimg-stuff into it.

This is already possible since version 8 if I don't make mistake. You can use `flameshot gui --raw` to output to stdout and pipe it to whatever you want. Some months ago I wrote a blog post in which I used this for reading barcode:

https://mehrad.ai/posts/20210702-extracting-payment-info-in-rasterized-invoice/

**Comment by xstable** [comment]
@mmahmoudian Ok, but I don't mean to create an own script which only use flameshot, I mean a section inside flameshot which do a predefined action based on the current made image.

For example there could be a button "A" for "Actions" and if you click on it, you can choose between the scripts you entered.

It's more handy in my opinion, as to have a shortcut on OS-Layer for each script, which finally use flameshot only internally.
This way you all the time only open flameshot, but you are able to pipe the made image to every of the 1000000 ;) scripts you made.

**Comment by mmahmoudian** [comment]
@xstable there have been active discussions about adding plugin support to Flameshot. If you are interested, I suggest searching them in the issues and discussions and also join our Matrix room to know the most recent discussions over there (#flameshot-org:matrix.org)

**Comment by stefanocudini** [comment]
I think this would be a great idea
![image](https://user-images.githubusercontent.com/47107/149207879-1f7145a3-3cac-441b-9afb-b6fa757fef4c.png)

**Comment by terminal-root** [comment]
A temporary solution for anyone eager for this functionality:

If you have `zbar` and `xclip` installed, you can set up a keybind for this one-liner:

```
flameshot gui -r | zbarimg -1q --raw - | xclip -selection clipboard
```

This copies the QR code link decoded from the capture to your clipboard.

**Comment by Scary-Guy** [comment]
I'm glad to see I'm not the only one who has had this idea, I'm just sorry it took me this long to think of it.

I'm also sad that no tools have implemented this yet that I have seen.

**Comment by arch-btw** [comment]
For those on wayland, this seems to work (slight modification to @deron-dev's command):

`flameshot gui -r | zbarimg -1q --raw - | wl-copy`

The scanned QR output will be copied to clipboard, just like the original command.

**Comment by sayore** [comment]
for the love of god PLEASE

**Comment by phulstaert** [comment]
There are a lot of wanted features that are possible through command line piping. 
- Reading a QR code
- Reading a 1D code
- Doing OCR
- special processing
- special uploading
- special scripts
- AI stuff
- ....

I would vote for a feature where we can create custom buttons to 'push the resulting image to a command-line prompt". This would make Flameshot exponentially more powerful for everyone.

**Comment by mmahmoudian** [comment]
@phulstaert 
> I would vote for a feature where we can create custom buttons to 'push the resulting image to a command-line prompt". This would make Flameshot exponentially more powerful for everyone.

This is true, and in principal it should be easy for a plugin to do this. Basically all Flameshot plugin system should do is to pass the image to the plugin. After that the plugin can store it in a temp file and `cat` the file into the chain of pipes the user has specified.

At the moment our plugin system (#3661) can somewhat do that, but it still has some rough edges and the PR is not yet merged.

**Comment by phulstaert** [comment]
@mmahmoudian 

I'm new at the repository, so sorry if I'm talking nonsense, but I don't totally agree with this.

A plugin could do a load of different stuff.
If you have a OCR or Barcode (1D/2D), you would expect there to be text in the clipboard, not an image.
If you would have a plugin to convert the image to svg, it would also be text, but with image/svg+xml (if the clipboard has a content-type sort of thing)
If you would have a prompt to push the image to an AI to make it cartoon-like, you would expect an image in the clipboard.

But if there are plugins, we should ask ourself: should flameshot check that the plugin does not return scripts or executable code (for security)?

And another thing is that plugins would have dependencies. How do we deal with this? Does it autoinstall them on debian and fedora distributions, of just a mention of them? How do we deal with errors (for example if a newer version of a dependancy is installed and not backwards compatible?)

And finally, not every plugin would be as fast as the current features. This would demand for a "please wait" 'modal', but maybe with a timeout if it should crash. Or -if possible- a progressbar if the plugin supports it.

I do agree that a quick and dirty implementation is possible, but for a professional product, it will take some time and thinking.
